lib/services/local/notifications.go

```diff
@@ -0,0 +1,298 @@
+	if err := services.ValidateUserNotificationState(state); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	state.Kind = types.KindUserNotificationState
+	state.Version = types.V1
+
+	if state.Metadata == nil {
+		state.Metadata = &headerv1.Metadata{}
+	}
+
+	// We set this to the notification UUID because the service adapter uses `getName()` to determine the backend key to use when storing the object.
+	state.Metadata.Name = state.Spec.NotificationId
+
+	// Append username prefix.
+	serviceWithPrefix, err := s.userNotificationStateService.WithPrefix(username)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	upserted, err := serviceWithPrefix.UpsertResource(ctx, state)
+	return upserted, trace.Wrap(err)
+}
+
+// DeleteUserNotificationState deletes a user notification state object.
+func (s *NotificationsService) DeleteUserNotificationState(ctx context.Context, username string, notificationId string) error {
+	// Append username prefix.
+	serviceWithPrefix, err := s.userNotificationStateService.WithPrefix(username)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = serviceWithPrefix.DeleteResource(ctx, notificationId)
+	return trace.Wrap(err)
+}
+
+// DeleteAllUserNotificationStatesForUser deletes all of a user's notification states.
+func (s *NotificationsService) DeleteAllUserNotificationStatesForUser(ctx context.Context, username string) error {
+	// Append username prefix.
+	serviceWithPrefix, err := s.userNotificationStateService.WithPrefix(username)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = serviceWithPrefix.DeleteAllResources(ctx)
+	return trace.Wrap(err)
+}
+
+// ListUserNotificationStates returns a page of a user's notification states.
+func (s *NotificationsService) ListUserNotificationStates(ctx context.Context, username string, pageSize int, nextToken string) ([]*notificationsv1.UserNotificationState, string, error) {
+	// Append username prefix.
+	serviceWithPrefix, err := s.userNotificationStateService.WithPrefix(username)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	states, nextToken, err := serviceWithPrefix.ListResources(ctx, pageSize, nextToken)
+	return states, nextToken, trace.Wrap(err)
+}
+
+// UpsertUserLastSeenNotification creates or updates a user's last seen notification item.
+func (s *NotificationsService) UpsertUserLastSeenNotification(ctx context.Context, username string, ulsn *notificationsv1.UserLastSeenNotification) (*notificationsv1.UserLastSeenNotification, error) {
+	if err := services.ValidateUserLastSeenNotification(ulsn); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ulsn.Kind = types.KindUserLastSeenNotification
+	ulsn.Version = types.V1
+
+	if ulsn.Metadata == nil {
+		ulsn.Metadata = &headerv1.Metadata{}
+	}
+	// We set this to the username because the service adapter uses `getName()` to determine the backend key to use when storing the object.
+	ulsn.Metadata.Name = username
```

Comment:
I don't think this is a concern since the `username` here is going to be populated with the name of the authenticated user, so we know that they exist. This is unlike the notification ID which a client can arbitrarily provide and thus should be verified as you suggested in your other comment.